Extract LinkedHashMap from LruCache

stepancheg opened the pull request:
LruCache is a thin wrapper around LinkedHashMap now.

LinkedHashMap now has iterator which is double-ended iterator.

fmt is implemented using iterator now.

Gankro commented:
Just like to say: yesssss

LinkedHashMap is my favourite data structure and now I can say Rust is a real programming language.

Gankro commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,540 @@
+
+//! A HashMap wrapper that holds key-value pairs in insertion order.
+//!
+//! # Example
```
Examples is the convention

stepancheg replied:
Ah, plural. I get it.

Gankro commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,540 @@
+/// A linked hash map.
+pub struct LinkedHashMap<K, V> {
+    map: HashMap<KeyRef<K>, Box<LinkedHashMapEntry<K, V>>>,
+    head: *mut LinkedHashMapEntry<K, V>,
```
Do we want to support a deque interface? (can be added in a later PR)

stepancheg replied:
Not sure about Deque. However, I'd like to implement `into_iter` one day (it is a somewhat harder than plain `iter`).

Gankro commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,540 @@
+
+impl<K: Hash + Eq, V> LinkedHashMap<K, V> {
+    /// Create a linked hash map.
+    #[unstable = "matches collection reform specification, waiting for dust to settle"]
```
Can probably ditch these markers.

Gankro commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,540 @@
+    }
+
+    /// Deprecated: Replaced with `insert`.
+    #[deprecated = "Replaced with `insert`"]
```
And the deprecated functionality.

stepancheg commented:
Updated PR with:

* better example, describing iteration order (is that what you requested?)
* removed all deprecated code
* removed all unstable markers

stepancheg commented:
Updated with plural "Examples" and minor example and test tweaks.

Gankro commented:
Unless you have an argument otherwise, I think the interface for interacting with the linked-ness should either be the stack (push/pop/peek) or deque (_front, _back) interface.

Gankro commented:
Although get_refresh is interesting...

Gankro commented on src/linked_hash_map.rs:
````diff
@@ -0,0 +1,509 @@
+    /// assert_eq!(map.get(&2), Some(&20));
+    /// ```
+    #[inline]
+    pub fn remove_lru(&mut self) {
````
I would call this `pop`

stepancheg replied:
`remove_front()` would be better IMHO. Because:
* iterator API uses `next_back` to return element from the opposite end
* somebody may want to have `remove_back`

Gankro replied:
convention from deque is pop_back. Usually `remove` involves giving some kind of key/index.

stepancheg replied:
So, it should be `pop_front`?

Gankro commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,509 @@
+
+    /// Return a value corresponding to the key in the map.
+    ///
+    /// If value is found, it is pushed to the end of the list.
```
I'm not sure `end` is a meaningful metaphor here? Are we providing a logical stack or queue? We should be using that terminology, I think.

stepancheg replied:
I think it is linked list. Not a queue, not a stack.

Gankro commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,509 @@
+    /// If value is found, it is pushed to the end of the list.
+    /// This operation can be used in implemenation of LRU cache.
+    ///
+    /// # Example
```
Examples

Gankro commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,509 @@
+    }
+
+    #[inline]
+    fn detach(&mut self, node: *mut LinkedHashMapEntry<K, V>) {
```
Might want to move these to a seperate `impl` block to seperate public/private spatially.

Gankro commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,509 @@
+    /// A double-ended iterator visiting all key-value pairs in order of insertion.
+    /// Iterator element type is `(&'a K, &'a V)`
+    ///
+    /// # Example
```
Examples, etc

Gankro commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,509 @@
+}
+
+impl<K: Hash + Eq, V> Extend<(K, V)> for LinkedHashMap<K, V> {
+    fn extend<T: Iterator<Item=(K, V)>>(&mut self, mut iter: T) {
```
Should `reserve` with size_hint here, I think.

stepancheg replied:
It is not safe, because iterator may contain billion pairs with the same key. Resulting map will have the same size as before and capacity of billion.

Gankro commented on src/lru_cache.rs:
```diff
@@ -100,16 +63,10 @@ impl<K: Hash + Eq, V> LruCache<K, V> {
-        let cache = LruCache {
-            map: HashMap::new(),
+        LruCache {
+            map: LinkedHashMap::new(),
```
Should this be with_capacity(capacity + 1), maybe?

stepancheg replied:
I think, it is safer to use default capacity, because LruCache might not be ever used at full capacity, not need to waste memory.

Gankro commented on src/lru_cache.rs:
```diff
@@ -343,19 +236,13 @@ impl<A: fmt::Show + Hash + Eq, B: fmt::Show> fmt::Show for LruCache<A, B> {
     /// used to least-recently used.
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
```
Can this not just call `fmt` on the LinkedHashMap?

stepancheg replied:
No, because `LruCache` uses reverse iterator in `fmt`, and `LinkedHashMap` uses forward iterator.

stepancheg commented:
Updated patch with issues fixed.

stepancheg commented:
Updated patch with `.keys()` and `.values()` iterators.

apasel422 commented on src/linked_hash_map.rs:
````diff
@@ -0,0 +1,594 @@
+//! ```rust
+//! use collect::LinkedHashMap;
+//!
+//! let mut map: LinkedHashMap<int, int> = LinkedHashMap::new();
````
I think this can just be
```rust
let mut map = LinkedHashMap::new();
```
now that integer fallback has landed.

stepancheg replied:
Updated patch without explicit type of LHM, however, I added type suffix like `1i`.

apasel422 replied:
You shouldn't need the suffixes now that integer fallback exists.

On Wed, Jan 7, 2015 at 3:15 PM, Stepan Koltsov <notifications@github.com>
wrote:

> In src/linked_hash_map.rs
> <https://github.com/Gankro/collect-rs/pull/66#discussion-diff-22612374>:
>
> > +//
> > +// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
> > +// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
> > +// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
> > +// option. This file may not be copied, modified, or distributed
> > +// except according to those terms.
> > +
> > +
> > +//! A HashMap wrapper that holds key-value pairs in insertion order.
> > +//!
> > +//! # Examples
> > +//!
> > +//! ```rust
> > +//! use collect::LinkedHashMap;
> > +//!
> > +//! let mut map: LinkedHashMap<int, int> = LinkedHashMap::new();
>
> Updated patch without explicit type of LHM, however, I added type suffix
> like 1i.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Gankro/collect-rs/pull/66/files#r22612374>.
>

stepancheg replied:
I'm not sure integer fallback won't take off again. (Personally I prefer no integer fallback than fallback to u32).

apasel422 commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,594 @@
+}
+
+impl<K: Hash + Eq, V> LinkedHashMap<K, V> {
+    /// Create a linked hash map.
```
s/Create/Creates/

apasel422 commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,594 @@
+    /// use collect::LinkedHashMap;
+    /// let mut map = LinkedHashMap::new();
+    ///
+    /// map.insert(1i, "a");
```
s/1i/1/

apasel422 commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,594 @@
+        old_val
+    }
+
+    /// Return a value corresponding to the key in the map.
```
s/Return/Returns/

(Same goes for the rest of the module.)

And should this be "_a_ value" or "_the_ value"?

apasel422 commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,594 @@
+    }
+}
+
+pub struct Iter<'a, K: 'a, V: 'a> {
```
Can you implement `ExactSizeIterator` for these iterators?

stepancheg replied:
Done, thanks.

apasel422 commented on src/linked_hash_map.rs:
```diff
@@ -0,0 +1,594 @@
+}
+
+/// A linked hash map.
+pub struct LinkedHashMap<K, V> {
```
Should this not be parameterized over the `Hasher`, like `HashMap` is?

stepancheg replied:
It should, but probably in another PR.

stepancheg commented:
Updated the patch with bunch of issues reported by @apasel422.

stepancheg commented:
Is anything else needed to merge this PR?

Gankro commented:
This LGTM, @apasel422 does this look good to you?

apasel422 commented:
@Gankro: Yup. I suspect changes in Rust master will break some things, but that can be addressed with a subsequent PR.

Gankro commented:
:boom: :heart:

stepancheg commented:
I'm trying to implement `iter_mut`, and I have trouble. Have a look at the code:

https://github.com/stepancheg/collect-rs/commit/edf6dae12346221c0c218d148621a43ec165428c#diff-9312c4317d6f1b3401488758a7a7a077R673

It should not work, because `iter_mut` mutably borrows map, and `map.get(..)` needs to borrow map again. But test passes.

Moreover siguatures of `struct IterMut` and `fn iter_mut` are identical to those in `HashMap`, however, if I replace `LinkedHashMap` with `HashMap`, compilation fails as I expect.

What have I missed?

Gankro commented:
You probably need to include a ContravariantLifetime for the `'a` in the struct.

stepancheg commented:
@Gankro that helps. Should this marker be also added to `Iter` struct? I see, `slice::Iter` has it.

Gankro commented:
Yeah, if a type has a lifetime that isn't used in the struct, you should use a marker so that it *is*.
